File: src/algorithms/binary-search-trees/min-height-bst/solution2.go
package main

// O(nlog(n)) time | O(n) space - where n is the length of the array
func MinHeightBST(array []int) *BST {
	return constructMinHeightBst(array, nil, 0, len(array)-1)
}

func constructMinHeightBst(array []int, bst *BST, startIdx, endIdx int) *BST {
	if endIdx < startIdx {
		return nil
	}
	midIdx := (startIdx + endIdx) / 2
	valueToAdd := array[midIdx]
	if bst == nil {
		bst = &BST{Value: valueToAdd}
	} else {
		bst.Insert(valueToAdd)
	}
	constructMinHeightBst(array, bst, startIdx, midIdx-1)
	constructMinHeightBst(array, bst, midIdx+1, endIdx)
	return bst
}

type BST struct {
	Value int

	Left *BST
	Right *BST
}

func (tree *BST) Insert(value int) *BST {
	if value < tree.Value {
		if tree.Left == nil {
			tree.Left = &BST{Value: value}
		} else {
			tree.Left.Insert(value)
		} 
	} else {
		if tree.Right == nil {
			tree.Right = &BST{Value: value}
		} else {
			tree.Right.Insert(value)
		}
	}
	return tree
}

File: src/algorithms/binary-search-trees/min-height-bst/solution2.js
// O(n) time | O(n) space 
// n = length of the array
function minHeightBst(array) {
  return constructMinHeightBst(array, null, 0, array.length - 1);
};

function constructMinHeightBst(array, bst, startIdx, endIdx) {
  if (endIdx < startIdx) return;
  const midIdx = Math.floor((startIdx + endIdx) / 2);
  const newBstNode = new BST(array[midIdx]);
  if (bst === null) {
    bst = newBstNode;
  } else {
    if (array[midIdx] < bst.value) {
      bst.left = newBstNode;
      bst = bst.left;
    } else {
      bst.right = newBstNode;
      bst = bst.right;
    };
  };
  constructMinHeightBst(array, bst, startIdx, midIdx - 1);
  constructMinHeightBst(array, bst, midIdx + 1, endIdx);
  return bst;
};

class BST {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  };

  insert(value) {
    if (value < this.value) {
      if (this.left === null) {
        this.left = new BST(value);
      } else {
        this.left.insert(value);
      };
    } else {
      if (this.right === null) {
        this.right = new BST(value);
      } else {
        this.right.insert(value);
      };
    };
  };
};

exports.minHeightBst = minHeightBst;